### PR TITLE
Issue #2131 - Introduce a monitored thread pool.

### PR DESCRIPTION
**Pull request by sbordet**
Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

### CODE REVIEW

**Comment by sbordet** [comment]
Need to decide whether to factor out the additional methods into an interface such as `ThreadPool.Monitored`, so that also `ExecutorThreadPool` can have its monitored version.

Need to implement the same for `ExecutorThreadPool`.

File: jetty-util/src/main/java/org/eclipse/jetty/util/thread/MonitoredQueuedThreadPool.java [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+import java.util.concurrent.atomic.LongAdder;
+
+import org.eclipse.jetty.util.Atomics;
+import org.eclipse.jetty.util.BlockingArrayQueue;
+import org.eclipse.jetty.util.annotation.ManagedAttribute;
+import org.eclipse.jetty.util.annotation.ManagedObject;
+import org.eclipse.jetty.util.annotation.ManagedOperation;
+
+@ManagedObject
+public class MonitoredQueuedThreadPool extends QueuedThreadPool
```
Comment (gregw) [inline jetty-util/src/main/java/org/eclipse/jetty/util/thread/MonitoredQueuedThreadPool.java]: No javadoc

File: jetty-util/src/main/java/org/eclipse/jetty/util/thread/MonitoredQueuedThreadPool.java [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+        {
+            @Override
+            public void run()
+            {
+                long queueLatency = System.nanoTime() - begin;
+                tasks.increment();
+                Atomics.updateMax(maxQueueLatency, queueLatency);
+                totalQueueLatency.add(queueLatency);
+                Atomics.updateMax(maxThreads, threads.incrementAndGet());
+                long start = System.nanoTime();
```
Comment (gregw) [inline jetty-util/src/main/java/org/eclipse/jetty/util/thread/MonitoredQueuedThreadPool.java]: Why not use our SampleStatistics class which gives your max, total and count?

File: jetty-util/src/main/java/org/eclipse/jetty/util/thread/MonitoredQueuedThreadPool.java [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+    public void execute(final Runnable job)
+    {
+        final long begin = System.nanoTime();
+        super.execute(new Runnable()
+        {
+            @Override
+            public void run()
+            {
+                long queueLatency = System.nanoTime() - begin;
+                tasks.increment();
```
Comment (gregw) [inline jetty-util/src/main/java/org/eclipse/jetty/util/thread/MonitoredQueuedThreadPool.java]: Why not use our CounterStatistics class which gives you current, max and total

File: jetty-util/src/main/java/org/eclipse/jetty/util/thread/MonitoredQueuedThreadPool.java [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+        return count == 0 ? -1 : totalTaskLatency.sum() / count;
+    }
+
+    @ManagedAttribute("the maximum task execution time, in nanoseconds")
+    public long getMaxTaskLatency()
+    {
+        return maxTaskLatency.get();
+    }
+
+    public static class MonitoringBlockingArrayQueue extends BlockingArrayQueue<Runnable>
```
Comment (gregw) [inline jetty-util/src/main/java/org/eclipse/jetty/util/thread/MonitoredQueuedThreadPool.java]: why is this needed?  If you use a CounterStatistics class for the tasks as they are queued and then run, you will already know the current and max size of the queue without double counting.

File: jetty-util/src/main/java/org/eclipse/jetty/util/thread/MonitoredQueuedThreadPool.java [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+                {
+                    job.run();
+                }
+                finally
+                {
+                    long taskLatency = System.nanoTime() - start;
+                    threads.decrementAndGet();
+                    Atomics.updateMax(maxTaskLatency, taskLatency);
+                    totalTaskLatency.add(taskLatency);
+                }
```
Comment (gregw) [inline jetty-util/src/main/java/org/eclipse/jetty/util/thread/MonitoredQueuedThreadPool.java]: I think it would also be interesting to set a thread local here, so we can measure how long after finishing one task before this thread picks up the next task.  If this is high it could indicate that some threads are being kept out from grabbing jobs by other busy threads... (I wonder if we are using fair locks?).   Anyway, set nanoTime into a thread local and check it on the next job.

Comment (sbordet) [reply jetty-util/src/main/java/org/eclipse/jetty/util/thread/MonitoredQueuedThreadPool.java]: If the server is idle, the value reported will be large not because threads are kept out, but just because there are no tasks - so I am not sure we will have any value out it this value.

We don't use fair locks because they are way slower than non-fair locks.

**Comment by sbordet** [comment]
@gregw I updated the implementation by using `CounterStatistics` and `SampleStatistics`.
We now record the total count of jobs multiple times (because each `*Statistics` instance does), we record values that we don't use, and the task statistics are less precise than before, but I can live with all of that.

File: jetty-util/src/main/java/org/eclipse/jetty/util/thread/MonitoredQueuedThreadPool.java (hunk 5 above)
Comment (gregw) [reply jetty-util/src/main/java/org/eclipse/jetty/util/thread/MonitoredQueuedThreadPool.java]: but if the server is busy and this value is still large, then than tells us something. But perhaps you are right and we don't need this until we suspect there is a fixable problem

File: jetty-util/src/main/java/org/eclipse/jetty/util/thread/MonitoredQueuedThreadPool.java [hunk 6]
```diff
@@ -18,28 +18,23 @@
...
-    private final LongAdder totalTaskLatency = new LongAdder();
-    private final MonitoringBlockingArrayQueue queue;
-    private final AtomicLong maxQueueLatency = new AtomicLong();
-    private final LongAdder totalQueueLatency = new LongAdder();
-    private final AtomicInteger threads = new AtomicInteger();
-    private final AtomicInteger maxThreads = new AtomicInteger();
+    private final CounterStatistic queueStats = new CounterStatistic();
+    private final SampleStatistic queueLatencyStats = new SampleStatistic();
+    private final SampleStatistic taskLatencyStats = new SampleStatistic();
+    private final CounterStatistic threadStats = new CounterStatistic();
```
Comment (gregw) [inline jetty-util/src/main/java/org/eclipse/jetty/util/thread/MonitoredQueuedThreadPool.java]: add these as beans and they will be added as JMX beans automagically.  No need to echo all their methods

Comment (gregw) [reply jetty-util/src/main/java/org/eclipse/jetty/util/thread/MonitoredQueuedThreadPool.java]: they will also get dumped

Comment (sbordet) [reply jetty-util/src/main/java/org/eclipse/jetty/util/thread/MonitoredQueuedThreadPool.java]: Will add them as beans, but neither `CounterStatistic` nor `SampleStatistic` are annotated with JMX annotations, so they will just show up as POJOs, which will only print their toString() during `dump()`. Their methods are not exposed anyway to JMX. 
The typical usage for the additional methods in `MonitoredThreadPool` such as `getMaxBusyThreads()` is actually from embedded usage during load testing, and I don't want to expose the `*Statistic` object.

Comment (sbordet) [reply jetty-util/src/main/java/org/eclipse/jetty/util/thread/MonitoredQueuedThreadPool.java]: Also, the `*Statistic` beans don't have a name to distinguish them so in the dump won't be possible to tell which bean represent what statistics without looking at the implementation.
Anyway, it's a cheap change and if we ever give the `*Statistic` beans a name they will already be dumped correctly.
